transcoder/__init__.py

```diff
@@ -17,26 +17,240 @@
+                 factory: str, schema_file_path: str, source_file_path: str, source_file_encoding: str,
+                 source_file_format_type: str, source_file_endian: str, prefix_length: int, skip_lines: int,
+                 skip_bytes: int,  message_skip_bytes: int, quiet: bool, output_type: str, output_encoding: str,
+                 output_path: str, error_output_path: str, destination_project_id: str, destination_dataset_id: str,
+                 message_handlers: str, lazy_create_resources: bool, frame_only: bool, stats_only: bool,
+                 create_schemas_only: bool, continue_on_error: bool, create_schema_enforcing_topics: bool,
+                 sampling_count: int, message_type_inclusions: str, message_type_exclusions: str, fix_header_tags: str,
+                 fix_separator: int, base64: bool, base64_urlsafe: bool):
+
+        signal.signal(signal.SIGINT, self.trap)
+
+        self.message_handler_spec = message_handlers
+        self.message_handlers = {}
+        self.all_message_type_handlers = []
+        self.all_handlers = []
+        self.handlers_enabled = False
+        self.continue_on_error = continue_on_error
+        self.error_output_path = error_output_path
+        self.output_encoding = output_encoding
+        self.frame_only = frame_only
+        self.create_schemas_only = create_schemas_only
+        self.lazy_create_resources = lazy_create_resources
+        self.prefix_length = prefix_length
+        self.quiet = quiet
+        self.start_time = None
+        self.stats_only = stats_only
+
+        self.output_prefix = os.path.basename(
+            os.path.splitext(source_file_path)[0]) if source_file_path else 'stdin'
+
+        self.error_writer = ErrorWriter(prefix=self.output_prefix,
+                                        output_path=self.error_output_path)
+
+        self.output_manager = get_output_manager(output_type, self.output_prefix, output_path,
+                                                output_encoding, self.prefix_length, destination_project_id,
+                                                destination_dataset_id, lazy_create_resources,
+                                                create_schema_enforcing_topics)
+
+        # TODO: think about this abstraction some more
+        if self.output_manager.supports_data_writing() is False:
+            self.create_schemas_only = True
+        else:
+            self.source = get_message_source(source_file_path, source_file_encoding,
+                                            source_file_format_type, source_file_endian,
+                                            skip_bytes, skip_lines, message_skip_bytes,
+                                            prefix_length, base64, base64_urlsafe)
+
+
+        if self.frame_only is False: # don't need a parser for just framing
+            self.message_parser: DatacastParser = get_message_parser(factory, schema_file_path,
```

Comment:
When this is True, the sampling_count should probably be supported for testing, etc. The fields that require cracking open the message should probably output a warning to the CLI to indicate that they cannot be used when framing only.

Reply:
Agree - let's say you only want to sample the first 5 messages unframed from the source and stop, this would come in handy there.

I think there's also a separate issue lurking where the transcoder doesn't halt processing once processed message count > sample_count.

Reply:
The reason for that is the sampling count was per message type. We don't know all message types that are in a file until we read the whole thing.

`        
if self.use_sampling is True and self.get_summary_count(message.name) >= self.sampling_count:
   message.ignored = True
   return message
`

Reply:
Good point, I wonder if it makes sense to broaden the sampling count logic and let users themselves narrow down which message types to in/exclude using `message_type_(in|ex)clusions` and/or `FilterHandler`s?

This would mean `--sampling_count 100` and no inclusions would just stream out the first 100 messages (of any type) then exit. As one narrows down the message universe with in/exclusions or `FilterHandler`s, the transcoder could just use the count of processed messages to evaluate against the specified threshold. 

For a message cracking warning under the frame only option, I was looking into mutually exclusive options with `argsparse` but that would be a relatively big change from the way options are being handled now. There are other areas where option mutexes would come in handy, though (like specifying `--output_path` when invoking with `--diag`.)

Will look for a way to express this in the code that's OK for the short term.

Reply:
Was able to restore this (globally, not per message) with a few lines of code. Keeping track of this logic is now in `Transcoder`, probably good to remove vestigial sample count references in `DatacastParser` as well.

Reply:
[c57944b](https://github.com/GoogleCloudPlatform/market-data-transcoder/pull/137/commits/c57944be89b265848b3a5ede3a2f3d40b7fff88d)